### DISCUSSION TITLE
Why does 'yarn install' rewrite package.json?

### DISCUSSION THREAD

**Discussion by rdicroce** [yarnpkg/berry#2636] [Q&A] [answered]
Like the title says: 'yarn install' causes package.json to be rewritten even though there's no reason Yarn needs to modify the file. I'm not sure if this is a bug, per se, but it's definitely surprising behavior. I wouldn't care, except that when Yarn writes package.json, it also sorts dependencies/peerDependencies/etc. So the file content after an install is semantically equivalent, but the raw bytes might be different.

As to why this is bothering me: I'm primarily a Java developer, and I'm accustomed to directly editing the package descriptor (pom.xml for Maven). I'm also accustomed to being able to group related dependencies together (in terms of how they're ordered in the file) and being able to leave comments in the descriptor to explain things, like why a dependency is needed, or why it's not on the latest version. Which you can't do with package.json, and I know I'm not the only person who finds that annoying, because some Googling finds plenty of people proposing various hacky solutions. There was even an issue about it here (#241) that was closed with no action taken because it's a complicated problem.

Well, I invented my own solution (actually the one mentioned in that issue): an automagic transformation between package.json and package.jsonc. JSON is the canonical file, and a JSONC is created from it. Comments can be added in the JSONC, and when it is saved, it gets transformed back to JSON, with the comments being parsed in a separate pass and added to the JSON as a valid JSON object. Similarly, if the JSON changes, it gets transformed to JSONC and the comments are re-inserted at the appropriate places.

It works pretty well, except for that problem of yarn rewriting the file and not preserving the ordering. I've made my magic smart enough that it still knows where the comments are supposed to go, but the re-ordering means I can't group things together.

So: should any of this be considered a bug in Yarn? IMO rewriting package.json for no reason feels like a bug. The sorting behavior also feels unintended, and ideally I'd like for Yarn to preserve the order, but I'd settle for not modifying the file unnecessarily.

**Comment by arcanis** [comment, marked as answer]
> So: should any of this be considered a bug in Yarn? IMO rewriting package.json for no reason feels like a bug. The sorting behavior also feels unintended, and ideally I'd like for Yarn to preserve the order, but I'd settle for not modifying the file unnecessarily.

Not keeping the dependencies in alphabetical order leads to harder-to-maintain projects as you lose the ability to quickly see which dependencies are in use. That they are enforced in a consistent order is intended 🙂

**Reply by rdicroce** [reply]
Okay, thanks for that info. Would you have any interest in a patch to allow that behavior to be disabled? I don't have time to do that right now, but maybe in the future.

**Reply by arcanis** [reply]
I'd incline towards no, I'm not convinced it's a behavior that's worth the (ever so small) extra complexity.

**Reply by jeffrson** [reply]
Well, as useful as ordering dependencies *can* be, OTOH, yarn removes "placeholders" like `"peerDependencies": {}` - what could be seen as "questionable".

**Reply by martinal** [reply]
Found this now because we have an issue with yarn install rewriting package.json with formatting that differs from what biome 1.7.1 wants to use, making the tools fight back and forth.

The diff was a silly
```
-  "workspaces": ["packages/agents"],
+  "workspaces": [
+    "packages/agents"
+  ],
```
where yarn wanted the multiline version.

Took me some time to debug and I had to disable formatting package.json by biome to get around it.

IMO it's not the role of yarn install to edit files that are commited to the repository, there are tools like biome for that.

In general I would expect that doing
```
git checkout ...
yarn install
git status
```
should not show modified files.

**Reply by IgnisDa** [reply]
@martinal Were you able to find a solution to this?

**Reply by Zhengqbbb** [reply]
🙄 Why rewrite `package.json.bin` and i can't disable the error behavior 👎

![CleanShot 2024-07-22 at 18 12 25@2x](https://github.com/user-attachments/assets/f68d90e6-1be9-4d56-b3a2-7a111d6addb6)

**Reply by vecerek** [reply]
This conflicts with my prettier setup and causes huge issues in my CI where prettier format checks are failing after `yarn install`. The ability to opt out of this behavior is important. Or at least the ability to configure prettier not to be at odds with yarn's ordering.

**Reply by pandeiro** [reply]
I'm doing `git reset --hard HEAD` after the `yarn install --immutable` step in my CI (which brings new meaning to the word 'immutable'! 😆) in order to work around this 🤷‍♂️

**Reply by hanseltime** [reply]
@arcanis - is there an issue with providing an opt-in flag for this?  (i.e. `yarn install --no-format`)

That would allow people who have more fine-grained repo toolsets (i.e. biome does these formatting checks, and yarn is just the installer), to opt-in to the functionality that yarn would otherwise provide for the less complex early project starter.

If so, I could potentially put together a PR.

**Reply by vilkinsons** [reply]
Just to chime in: something like the solution @hanseltime offers up above would be [used](https://github.com/hashintel/hash/pull/6031#issuecomment-2565963601) and appreciated from our side, as well, if @arcanis might be open to its inclusion. In the meantime, thank you for all the ongoing hard work, and a happy new year!

**Reply by fbartho** [reply]
I do keep my dependencies in alphabetical order -- and yet this is still a problem.

- Rewriting my package.json and removing the occasional, intentional, blank lines and arbitrarily reformatting it does not match the standards of every other JSON File in my repository.
- Additionally, arbitrarily rewriting package.json can trigger `watchman` and other directory-monitoring tools. This should only happen if there's a logical difference, and rewriting `package.json` could be an indication that you need to revalidate `yarn.lock` and relaunch other processes.

So this accepted answer perhaps discards the use-case of the user, but does not address the problem that **yarn shouldn't do this**

**Reply by natew** [reply]
No format is fine, I don't mind the sorting so long as it keeps the same whitespace tbh. I have to disable formatting of package.json file specifically in my editor which is annoying.

**Reply by hanseltime** [reply]
After having been bit by this a number of times, I went ahead and wrote a plugin that solves this issue reliably.

https://github.com/HanseltimeIndustries/correct-yarn-formatting

It will rewrite any `package.json` changes that did not result in  change to the actual package structure and works in monorepos.  

Feel free to give it a try until this maybe becomes an internal feature!

**Reply by ColtHands** [reply]
Does `pnpm` have this issue?

**Reply by pastelmind** [reply]
> Does `pnpm` have this issue?

Nope. `pnpm install` does not modify `package.json`.

**Comment by fbartho** [comment]
I've filed a Bug ticket here: https://github.com/yarnpkg/berry/issues/6844

The discussion has successfully answered the question asking "why?", but the behavior is still indicative of undesired behavior from first principles!
